src/types/mod.rs

```diff
@@ -7,7 +7,321 @@
+        fast_scan = |$scanner:ident| $fast_scan:expr;
+        format = |&$self:ident, $formatter:ident| $format:expr;
+    }) => {
+        impl $type {
+            pub const TAG: Tag = Tag::new($tag);
+
+            pub fn fast_scan(scanner: &mut FastScanner<'_>) -> Result<Self, FastScanError> {
+                Self::fast_scan_inner(scanner.scan_struct(Self::TAG)?)
+            }
+
+            pub fn fast_scan_with(scanner: &mut FastScanner<'_>, tag: Tag) -> Result<Self, FastScanError> {
+                Self::fast_scan_inner(scanner.scan_struct(tag)?)
+            }
+
+            pub fn fast_scan_opt(scanner: &mut FastScanner<'_>) -> Result<Option<Self>, FastScanError> {
+                scanner
+                    .scan_opt_struct(Self::TAG)?
+                    .map(Self::fast_scan_inner)
+                    .transpose()
+            }
+
+            pub fn fast_scan_opt_with(scanner: &mut FastScanner<'_>, tag: Tag) -> Result<Option<Self>, FastScanError> {
+                scanner.scan_opt_struct(tag)?.map(Self::fast_scan_inner).transpose()
+            }
+
+            fn fast_scan_inner(mut $scanner: FastScanner<'_>) -> Result<Self, FastScanError> {
+                let this = $fast_scan;
+                $scanner.finish()?;
+                Ok(this)
+            }
+
+            pub fn format(&self, formatter: &mut Formatter<'_>) -> FormatResult {
+                self.format_with(formatter, Self::TAG)
+            }
+
+            pub fn format_with(&$self, formatter: &mut Formatter<'_>, tag: Tag) -> FormatResult {
+                let mut $formatter = formatter.format_struct(tag)?;
+                $format;
+                Ok($formatter.finish())
+            }
+        }
+    };
+
+    // Impl TTLV (de)serialization for a boolean wrapper.
+    (bool $type:ident as $tag:literal) => {
+        impl $type {
+            pub const TAG: Tag = Tag::new($tag);
+
+            pub fn fast_scan(scanner: &mut FastScanner<'_>) -> Result<Self, FastScanError> {
+                Self::fast_scan_with(scanner, Self::TAG)
+            }
+
+            pub fn fast_scan_with(scanner: &mut FastScanner<'_>, tag: Tag) -> Result<Self, FastScanError> {
+                scanner.scan_bool(tag).map(Self)
+            }
+
+            pub fn fast_scan_opt(scanner: &mut FastScanner<'_>) -> Result<Option<Self>, FastScanError> {
+                Self::fast_scan_opt_with(scanner, Self::TAG)
+            }
+
+            pub fn fast_scan_opt_with(scanner: &mut FastScanner<'_>, tag: Tag) -> Result<Option<Self>, FastScanError> {
+                scanner.scan_opt_bool(tag).map(|s| s.map(Self))
+            }
+
+            pub fn format(&self, formatter: &mut Formatter<'_>) -> FormatResult {
+                self.format_with(formatter, Self::TAG)
+            }
+
+            pub fn format_with(&self, formatter: &mut Formatter<'_>, tag: Tag) -> FormatResult {
+                formatter.format_bool(tag, self.0)
+            }
+        }
+    };
```

Comment:
Could this impl be simplified by invoking `impl_serde_ttlv!` again, something like:

```suggestion
    (bool $type:ident as $tag:literal) => {
        impl_ttlv_serde!(struct $type as $tag {
          fast_scan = |scanner| scanner.scan_bool($tag).map(Self),
          format = |&self, formatter| formatter.format_bool($tag, self.0)
        }
    };
```

Reply:
No, because the `struct` impl calls `.scan_struct()` and `.format_struct()` for you.